component/loki/source/kubernetes_events/event_controller.go

```diff
@@ -0,0 +1,306 @@
+			AddFunc:    func(obj interface{}) { ctrl.onAdd(ctx, obj) },
+			UpdateFunc: func(oldObj, newObj interface{}) { ctrl.onUpdate(ctx, oldObj, newObj) },
+			DeleteFunc: func(obj interface{}) { ctrl.onDelete(ctx, obj) },
+		})
+	}
+	return nil
+}
+
+func (ctrl *eventController) onAdd(ctx context.Context, obj interface{}) {
+	event, ok := obj.(*corev1.Event)
+	if !ok {
+		level.Warn(ctrl.log).Log("msg", "received an event for a non-Event Kind", "type", fmt.Sprintf("%T", obj))
+		return
+	}
+	err := ctrl.handleEvent(ctx, event)
+	if err != nil {
+		level.Error(ctrl.log).Log("msg", "error handling event", "err", err)
+	}
+}
+
+func (ctrl *eventController) onUpdate(ctx context.Context, oldObj, newObj interface{}) {
+	oldEvent, ok := oldObj.(*corev1.Event)
+	if !ok {
+		level.Warn(ctrl.log).Log("msg", "received an event for a non-Event Kind", "type", fmt.Sprintf("%T", oldObj))
+		return
+	}
+	newEvent, ok := newObj.(*corev1.Event)
+	if !ok {
+		level.Warn(ctrl.log).Log("msg", "received an event for a non-Event Kind", "type", fmt.Sprintf("%T", newObj))
+		return
+	}
+
+	if oldEvent.GetResourceVersion() == newEvent.GetResourceVersion() {
+		level.Debug(ctrl.log).Log("msg", "resource version didn't change, ignoring call to onUpdate", "resource version", newEvent.GetResourceVersion())
+		return
+	}
+
+	err := ctrl.handleEvent(ctx, newEvent)
+	if err != nil {
+		level.Error(ctrl.log).Log("msg", "error handling event", "err", err)
+	}
+}
+
+func (ctrl *eventController) onDelete(ctx context.Context, obj interface{}) {
+	// no-op: the event got deleted from Kubernetes, but there's nothing to log
+	// when this happens.
+}
+
+func (ctrl *eventController) handleEvent(ctx context.Context, event *corev1.Event) error {
+	eventTs := eventTimestamp(event)
+
+	// Events don't have any ordering guarantees, so we can't rely on comparing the timestamp of
+	// this event to any other event received.
+	//
+	// We use a best-effort attempt to not re-deliver any events we've already
+	// logged by checking the timestamp from when the worker started. This may
+	// still cause us to drop some events in between recreating workers, but it
+	// minimizes risk.
+	//
+	// TODO(rfratto): a longer term solution would be to track timestamps for
+	// each involved object (or something similar), but that solution would need
+	// to make sure to not leak those timestamps, and find a way to recognize
+	// that involved objects have been deleted.
+	if !eventTs.After(ctrl.initTimestamp) {
+		return nil
+	}
+
+	lset, msg, err := ctrl.parseEvent(event)
+	if err != nil {
+		return err
+	}
+
+	entry := loki.Entry{
+		Entry: logproto.Entry{
+			Timestamp: eventTs,
+			Line:      msg,
+		},
+		Labels: lset,
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case ctrl.handler.Chan() <- entry:
+		// Update position offset only after it's been sent to the next set of
+		// components.
+		ctrl.task.Positions.Put(ctrl.positionsKey, "", eventTs.UnixMicro())
```

Comment:
Should we also update initTimestamp here? Currently it's being read from the positions file only on task startup.

Reply:
That's on purpose :) From a comment at the start of handleEvent:

> We use a best-effort attempt to not re-deliver any events we've already logged by checking the timestamp from when the worker started. 

Events don't have a guaranteed order, so updating initPosition as we receive events could cause us to drop many, many more events.

Reply:
Oh, I see!